source/grid/tria_accessor.cc

```diff
@@ -2845,368 +2845,432 @@ CellAccessor<dim, spacedim>::neighbor_child_on_subface(
     {
       case 2:
         {
-          const unsigned int neighbor_neighbor =
-            this->neighbor_of_neighbor(face);
-          const unsigned int neighbor_child_index =
-            GeometryInfo<dim>::child_cell_on_face(
-              this->neighbor(face)->refinement_case(),
-              neighbor_neighbor,
-              subface);
-
-          TriaIterator<CellAccessor<dim, spacedim>> sub_neighbor =
-            this->neighbor(face)->child(neighbor_child_index);
-          // the neighbors child can have children,
-          // which are not further refined along the
-          // face under consideration. as we are
-          // normally interested in one of this
-          // child's child, search for the right one.
-          while (sub_neighbor->has_children())
+          if (this->reference_cell_type() == ReferenceCell::Type::Tri)
             {
-              Assert((GeometryInfo<dim>::face_refinement_case(
-                        sub_neighbor->refinement_case(), neighbor_neighbor) ==
-                      RefinementCase<dim>::no_refinement),
-                     ExcInternalError());
-              sub_neighbor =
-                sub_neighbor->child(GeometryInfo<dim>::child_cell_on_face(
-                  sub_neighbor->refinement_case(), neighbor_neighbor, 0));
+              // only for isotropic refinement at the moment
+              Assert(this->neighbor(face)->refinement_case() ==
+                       RefinementCase<2>::isotropic_refinement,
+                     ExcNotImplemented());
+
+              // indices of child cells that are adjacent to a certain face of
+              // the mother cell
+              // basically GeometryInfo<2>::child_cell_on_face() for triangles
+              constexpr unsigned int faces_per_cell = 3;
+              constexpr unsigned int
+                subcells[faces_per_cell]
+                        [GeometryInfo<2>::max_children_per_face] = {{0, 1},
+                                                                    {1, 2},
+                                                                    {2, 0}};
```

Comment:
This information should be move to `ReferenceCell::internal::Info` (the new `GeometryInfo`).